### PR TITLE
[Offline mode]: Add web component to inform user about new data

### PR DESCRIPTION
**Pull request by tkohr**
PR refactors the logic of the current offline web component into a service and adds a second web component to "alert" the user on offline data updates.

Todos:

- [ ] Test on mobile

### CODE REVIEW

**Review by fgravin — COMMENTED** [review]
Ok, I don't have much to say cause I didn't really looked at the logic of the component.
Still some inline comments.

I think that ideally, such service should be a singleton, so you have to export the instance of the class and use it in your components.

The alert component is not well structured, I would have a generic alert component, and a smart component for the offline which is responsible to send the content of the alert to the alert component.

Lack of typing and confusing naming.

But overall yes, a singleton which manages its own state through Observables with components or other services subscribing to them is a good practice !

+ additional pure state services to glue the overall logic.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-alert.component.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  @state()
+  private status;
+
+  @query('.modal')
+  private modal: HTMLElement;
+
+  constructor() {
+    super();
+    this.offlineService = new LuxOfflineService();
+    this.offlineService.status$.subscribe((status)=> {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-alert.component.ts]: oh yes, I added it on both components

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.component.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    private menuDisplayed;
+
+    @state()
+    private status;
+
+    private offlineService: LuxOfflineService;
+
+    constructor() {
+        super();
+        this.offlineService = new LuxOfflineService();
```
[the remark this answers is not in the document]
Comment (tkohr) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.component.ts]: thanks, I changed it into a singleton

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.service.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    fetch(this.baseURL + "/check")
+      .then((response) => response.json())
+      .then((statusJson) => this.setStatus(statusJson))
+      .catch((error) => {
+        console.error('Error:', error);
+      });
+  }
+
+  private setStatus(tiles) {
+    this.tilePackages = {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.service.ts:35]: is only assigned here now

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.service.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  public hasLocalServer(){
+    return !!this.server;
+  }
+
+  private checkTiles() {
+    fetch(this.baseURL + "/check")
+      .then((response) => response.json())
+      .then((statusJson) => this.setStatus(statusJson))
```
[the remark this answers is not in the document]
Comment (tkohr) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.service.ts:28]: true, I adapted the naming

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-alert.component.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      if (status === 'UP_TO_DATE') {
+        $(this.modal).modal('hide');
+      }
+      this.status = status;
+    });
+  }
+
+  render() {
+    return html`
+      <div class="modal" tabindex="-1" role="dialog">
```
[the remark this answers is not in the document]
Comment (tkohr) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-alert.component.ts:43]: good point, tried to take the UI part out into a generic modal component, but had some troubles with the change detection. I will try to add this in a separate PR

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.component.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        this.offlineService.status$.subscribe((status)=> {
+          this.status = status;
```
[the remark this answers is not in the document]
Comment (tkohr) [reply geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline.component.ts:29]: hum, I don't know

**Comment by tkohr** [comment]
Thanks for the review @fgravin. I tried to address most points and rebased the PR.

**Review by mki-c2c — APPROVED** [review]
works very well and code is nicely structured

I added the dialog to refresh the page after offline download

**Comment by mki-c2c** [comment]
@tkohr could have a look at the last commit, if that looks OK to you ? (https://github.com/Geoportail-Luxembourg/geoportailv3/pull/2916/commits/f21f6af6fe52fde9b9dd78bb8db8f547650ab45e)

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-reload.component.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+  constructor() {
+    super();
+    //this.offlineService = new LuxOfflineService();
+    this.prevStatus = LuxOfflineServiceInstance.status$.getValue();
+    LuxOfflineServiceInstance.status$.subscribe((status)=> {
+      if (status != 'IN_PROGRESS' && this.prevStatus != undefined
+          && this.prevStatus != 'NO_INIT'
+        && status != this.prevStatus) {
+        $(this.modal).modal('show');
```
Comment (tkohr) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-reload.component.ts:28]: Uuh, not easy to follow this condition, but I guess it's correct. I'd use the `enum` values and strict checking (`!==`) anyhow.

File: geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-reload.component.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  render() {
+    return html`
+      <div class="modal" tabindex="-1" role="dialog">
+        <div class="modal-dialog offline-modal" role="document">
+          <div class="modal-content">
+            <div class="modal-header">
+              <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+              <h4 class="modal-title">${i18next.t('App reload needed: Offline data updated successfully')}</h4>
```
Comment (tkohr) [inline geoportal/geoportailv3_geoportal/static-ngeo/js/offline/lux-offline/lux-offline-reload.component.ts:42]: I think we still have to extract/translate the new strings (2nd step in https://github.com/Geoportail-Luxembourg/geoportailv3/tree/integration#translations).

**Review by tkohr — COMMENTED** [review]
Thanks a lot for finishing up this long awaiting PR with this missing part @mki-c2c! 
Code LGTM, I just left some minor comments.

**Pull request merged**